File: frontend/src/types/compute.ts
// ─── Job ──────────────────────────────────────────────────────────────────────

export type JobStatus =
  | 'pending'
  | 'splitting'
  | 'processing'
  | 'validating'
  | 'completed'
  | 'failed'

export type ChunkStatus = 'pending' | 'assigned' | 'done' | 'rejected'

export interface Job {
  id: string
  client_id: string
  job_type: string
  status: JobStatus
  params: Record<string, unknown>
  total_chunks: number
  completed_chunks: number
  reward_total: number
  result: Record<string, unknown> | null
  created_at: string       // ISO 8601 UTC
  completed_at: string | null
  progress: number         // 0–100, calculado por el backend
}

export interface JobListResponse {
  count: number
  jobs: Job[]
}

// ─── Chunks ───────────────────────────────────────────────────────────────────

export interface ChunkWithPayload {
  chunk_id: string
  job_id: string
  chunk_index: number
  job_type: string
  payload: Record<string, unknown>
}

export interface ClaimResponse {
  chunks: ChunkWithPayload[]
}

// ─── Submit ───────────────────────────────────────────────────────────────────

export interface SubmitRequest {
  result: Record<string, unknown>
  duration_ms: number
}

export interface SubmitResponse {
  chunk_result_id: string
  chunk_id: string
  status: ChunkStatus
  message: string
}

// ─── Create ───────────────────────────────────────────────────────────────────

export type JobOperation = 'mean' | 'sum' | 'min' | 'max' | 'count'

export interface JobCreateRequest {
  job_type: 'data-processing'
  params: {
    operation: JobOperation
    columns?: string[]
  }
}

// ─── Result ───────────────────────────────────────────────────────────────────

export interface JobResultResponse {
  id: string
  status: JobStatus
  result: Record<string, unknown>
  total_chunks: number
  completed_chunks: number
  completed_at: string
}

File: backend/app/models/compute.py
"""
Pydantic models for the compute pipeline (jobs, chunks, work endpoints).
"""
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ── Jobs ──────────────────────────────────────────────────────────────────────

class JobCreateRequest(BaseModel):
    job_type: str = Field(..., pattern="^data-processing$")
    params: dict[str, Any]


class JobPublic(BaseModel):
    id: UUID
    client_id: UUID
    job_type: str
    status: str
    params: dict[str, Any]
    total_chunks: int
    completed_chunks: int
    reward_total: float
    result: dict[str, Any] | None
    created_at: datetime
    completed_at: datetime | None
    progress: float  # completed_chunks / total_chunks * 100, computed by backend

    model_config = ConfigDict(from_attributes=True)


class JobListResponse(BaseModel):
    count: int
    jobs: list[JobPublic]


class JobResultResponse(BaseModel):
    id: UUID
    status: str
    result: dict[str, Any]
    total_chunks: int
    completed_chunks: int
    completed_at: datetime | None


# ── Chunks ────────────────────────────────────────────────────────────────────

class ChunkWithPayload(BaseModel):
    chunk_id: UUID
    job_id: UUID
    chunk_index: int
    job_type: str
    payload: dict[str, Any]


class ClaimRequest(BaseModel):
    max_chunks: int = Field(default=3, ge=1, le=10)


class ClaimResponse(BaseModel):
    chunks: list[ChunkWithPayload]


# ── Submit ────────────────────────────────────────────────────────────────────

class SubmitRequest(BaseModel):
    result: dict[str, Any]
    duration_ms: int = Field(..., gt=0)


class SubmitResponse(BaseModel):
    chunk_result_id: UUID
    chunk_id: UUID
    status: str   # chunk status after submit: "assigned" | "done" | "rejected"
    message: str
